nodebuilder/blobstream/service.go

```diff
@@ -0,0 +1,537 @@
+	}
+	_, proofs := merkle.ProofsFromByteSlices(dataRootEncodedTuples)
+	return proofs[height-int64(tuples[0].height)], nil
+}
+
+// fetchDataRootTuples takes an end exclusive range of heights and fetches its
+// corresponding data root tuples.
+func (s *Service) fetchDataRootTuples(ctx context.Context, start, end uint64) ([]DataRootTuple, error) {
+	tuples := make([]DataRootTuple, 0, end-start)
+	for height := start; height < end; height++ {
+		block, err := s.headerServ.GetByHeight(ctx, height)
+		if err != nil {
+			return nil, err
+		}
+		if block == nil {
+			return nil, fmt.Errorf("couldn't load block %d", height)
+		}
+		tuples = append(tuples, DataRootTuple{
+			height:   block.Height(),
+			dataRoot: *(*[32]byte)(block.DataHash),
+		})
+	}
+	return tuples, nil
+}
+
+// ProveShares generates a share proof for a share range.
+// Note: queries the whole EDS to generate the proof.
+// This can be improved by selecting the set of shares that will need to be used to create
+// the proof and only querying them. However, that would require re-implementing the logic
+// in Core. Also, core also queries the whole EDS to generate the proof. So, it's fine for
+// now. In the future, when blocks get way bigger, we should revisit this and improve it.
+func (s *Service) ProveShares(ctx context.Context, height uint64, start, end uint64) (*ResultShareProof, error) {
+	log.Debugw("proving share range", "start", start, "end", end, "height", height)
+	if height == 0 {
+		return nil, fmt.Errorf("height cannot be equal to 0")
+	}
+	if start == end {
+		return nil, fmt.Errorf("start share cannot be equal to end share")
+	}
+	if start > end {
+		return nil, fmt.Errorf("start share %d cannot be greater than end share %d", start, end)
+	}
+
+	log.Debugw("getting extended header", "height", height)
+	extendedHeader, err := s.headerServ.GetByHeight(ctx, height)
+	if err != nil {
+		return nil, err
+	}
+	log.Debugw("getting eds", "height", height)
+	eds, err := s.shareServ.GetEDS(ctx, extendedHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	startInt, err := uint64ToInt(start)
+	if err != nil {
+		return nil, err
+	}
+	endInt, err := uint64ToInt(end)
+	if err != nil {
+		return nil, err
+	}
+	odsShares, err := shares.FromBytes(eds.FlattenedODS())
+	if err != nil {
+		return nil, err
+	}
+	nID, err := pkgproof.ParseNamespace(odsShares, startInt, endInt)
+	if err != nil {
+		return nil, err
+	}
+	log.Debugw("generating the share proof", "start", start, "end", end, "height", height)
+	proof, err := pkgproof.NewShareInclusionProofFromEDS(eds, nID, shares.NewRange(startInt, endInt))
+	if err != nil {
+		return nil, err
+	}
+	return &ResultShareProof{ShareProof: proof}, nil
+}
+
+// ProveCommitment generates a commitment proof for a share commitment.
+// It takes as argument the height of the block containing the blob of data, its
+// namespace and its share commitment.
+// Note: queries the whole EDS to generate the proof.
```

Comment:
similar, should we create an issue for this to track this improvement?

Reply:
does this api essentially supersede the `blob.GetProof` api? I don't really see a need for it to exist if this is delivered. However, I'm unsure if `blobstream` if focused on some eth integration stuff or this is meant to be a general purpose proof api (limited understanding here, apologies!)

Reply:
`blob.GetProof` returns a different proof. It returns the proof of the namespace in the rows that contain the blob, which is proving shares >= the shares of the blob.

Reply:
The GetProof will stay as a canonical way to serve proofs for blobs long term. The GetProof is supposed to return the proof of just the blob, rather than the whole namespace. You can consider current behaviour a bug.

Reply:
Fetching whole EDS is okay short term solution and warrants an issue.

Reply:
Yes, that's the goal. The DataAvailbailityHeader is a detail of Celestia protocol that should have been abstracted away; similarly, the proofs our API provides should not prove to it, but to a single DataRoot hash.

Reply:
@rach-id, agreed, let's make this in a separate PR, thank you!

Reply:
https://github.com/celestiaorg/celestia-node/issues/3516